src/toil/server/wes/toil_backend.py

```diff
@@ -0,0 +1,318 @@
+                            task_id=self.run_id,  # set the Celery task ID the same as our run ID
+                            ignore_result=True)
+
+    def get_output_files(self) -> Any:
+        """
+        Return a collection of output files that this workflow generated.
+        """
+        return json.loads(self.fetch("outputs.json", "{}"))
+
+
+class ToilBackend(WESBackend):
+    """
+    WES backend implemented for Toil to run CWL, WDL, or Toil workflows. This
+    class is responsible for validating and executing submitted workflows.
+
+    Single machine implementation -
+    Use Celery as the task queue and interact with the "workflows/" directory
+    in the filesystem to store and retrieve data associated with the runs.
+    """
+
+    def __init__(self, work_dir: str, options: List[str]) -> None:
+        super(ToilBackend, self).__init__(options)
+        self.work_dir = work_dir
+        self.supported_versions: Dict[str, List[str]] = {}
+
+    def register_wf_type(self, name: str, supported_versions: List[str]) -> None:
+        """
+        Register a workflow type that this backend should support.
+        """
+        self.supported_versions[name.lower()] = supported_versions
+
+    def _get_run(self, run_id: str, should_exists: Optional[bool] = None) -> ToilWorkflow:
+        """
+        Helper method to instantiate a ToilWorkflow object.
+
+        :param run_id: The run ID.
+        :param should_exists: If set, ensures that the workflow run exists (or
+                              does not exist) according to the value.
+        """
+        run = ToilWorkflow(run_id, work_dir=os.path.join(self.work_dir, run_id))
+
+        if should_exists is not None:
+            if should_exists and not run.exists():
+                raise WorkflowNotFoundException
+            if should_exists is False and run.exists():
+                raise WorkflowConflictException(run_id)
+
+        return run
+
+    def get_runs(self) -> Generator[Tuple[str, str], None, None]:
+        """ A generator of a list of run ids and their state."""
+        if not os.path.exists(self.work_dir):
+            return
+
+        for run_id in os.listdir(self.work_dir):
+            run = self._get_run(run_id)
+            if run.exists():
+                yield run_id, run.get_state()
+
+    def get_state(self, run_id: str) -> str:
+        """
+        Return the state of the workflow run with the given run ID. May raise
+        an error if the workflow does not exist.
+        """
+        return self._get_run(run_id, should_exists=True).get_state()
+
+    @handle_errors
+    def get_service_info(self) -> Dict[str, Any]:
+        """ Get information about the Workflow Execution Service."""
+
+        state_counts = {state: 0 for state in (
+            "QUEUED", "INITIALIZING", "RUNNING", "COMPLETE", "EXECUTOR_ERROR", "SYSTEM_ERROR", "CANCELING", "CANCELED"
+        )}
+        for _, state in self.get_runs():
+            state_counts[state] += 1
```

Comment:
If we do ever get a partial read of the name of a state, this will crash because it will look for an entry that isn't there.
Also, maybe state_counts would be simpler as a collections.Counter? Do we need the entries for states that nothing is in?

Reply:
It doesn't say so in the spec, but I assume we don't need to include them? I'm going to use a `Counter` for now.